src/machine/flash.go

```diff
@@ -89,38 +93,50 @@ func (fl *FlashBuffer) Read(p []byte) (n int, err error) {
 	return
 }
 
-// Write data to a FlashBuffer.
+// Write data to a FlashBuffer. You must pass in a slice containing a multiple
+// of the WriteBlockSize() for the target device, or the data written will be
+// padded to fill in any missing space.
 func (fl *FlashBuffer) Write(p []byte) (n int, err error) {
-	// any new pages needed?
+	// any new blocks needed?
 	// NOTE probably will not work as expected if you try to write over page boundary
 	// of pages with different sizes.
-	pagesize := uintptr(fl.b.EraseBlockSize())
+	blocksize := uintptr(fl.b.EraseBlockSize())
 
 	// calculate currentPageBlock relative to fl.start, meaning that
 	// block 0 -> fl.start
-	// block 1 -> fl.start + pagesize
-	// block 2 -> fl.start + pagesize*2
+	// block 1 -> fl.start + blocksize
+	// block 2 -> fl.start + blocksize*2
 	// ...
-	currentPageBlock := (fl.start + fl.offset - FlashDataStart()) + (pagesize-1)/pagesize
-	lastPageBlockNeeded := (fl.start + fl.offset + uintptr(len(p)) - FlashDataStart()) + (pagesize-1)/pagesize
+	currentPageBlock := (fl.offset + (blocksize - 1)) / blocksize
+	lastPageBlockNeeded := (fl.offset + uintptr(len(p)) + (blocksize - 1)) / blocksize
+
+	if fl.erased == 0 {
+		// since we are writing for the first time, we need to
+		// erase the first page e.g. page 0
+		currentPageBlock = 0
+		lastPageBlockNeeded = 1
+	}
 
 	// erase enough blocks to hold the data
-	if err := fl.b.EraseBlocks(int64(currentPageBlock), int64(lastPageBlockNeeded-currentPageBlock)); err != nil {
-		return 0, err
+	if lastPageBlockNeeded*blocksize > fl.erased {
+		if err := fl.b.EraseBlocks(int64(currentPageBlock), int64(lastPageBlockNeeded-currentPageBlock)); err != nil {
+			return 0, err
+		}
+		fl.erased = lastPageBlockNeeded * blocksize
```

Comment:
I believe this approach will not work.
For example, when using the filesystem, the areas are not used in order from the beginning to the end.
It is also very confusing in combination with Seek.

I guess the only way is to read, erase, rewrite, and then write.

Reply:
> I guess the only way is to read, erase, rewrite, and then write.

I think that is probably true when using the `FlashBuffer`. It is just a shortcut? Perhaps it is not really complete enough as @aykevl suggests we should remove it.

For a real file system, try out https://github.com/tinygo-org/tinyfs/tree/dev which uses the `BlockDevice` interface.

Reply:
If so, how about the following?
`EraseBlocks()` are a little more difficult to handle.
And even if not necessarily using tinyfs, there may be cases where the user wants to save a little data.

* Write does not delete, but simply writes
* Read simply reads
* Create a function that returns the block number contained in the specified address to assist EraseBlock
    * ex) func XXXXX(addr uintptr) (blockNo, startAddr)